include/albatross/src/stats/chi_squared.hpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2019 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_STATS_CHI_SQUARED_HPP_
+#define ALBATROSS_STATS_CHI_SQUARED_HPP_
+
+/*
+ * Methods for computing quantities related to the Chi-Squared distribution
+ *
+ * https://en.wikipedia.org/wiki/Chi-squared_distribution
+ *
+ * Based largely off the implementation in the Stats C++ package:
+ *
+ * https://github.com/kthohr/stats/blob/master/include/stats_incl/prob/pchisq.ipp
+ */
+namespace albatross {
+
+namespace details {
+
+inline double chi_squared_cdf_unsafe(double x, std::size_t degrees_of_freedom) {
+  return incomplete_gamma(0.5 * degrees_of_freedom, 0.5 * x);
+}
+
+inline double chi_squared_cdf_safe(double x, std::size_t degrees_of_freedom) {
+
+  if (std::isnan(x)) {
+    return NAN;
+  }
+
+  if (std::numeric_limits<double>::epsilon() > x) {
+    return 0.;
+  }
+
+  if (degrees_of_freedom == 0) {
+    return 1.;
+  }
+
+  if (std::isinf(x)) {
+    return 1.;
+  }
+
+  return chi_squared_cdf_unsafe(x, degrees_of_freedom);
+}
+
+} // namespace details
+
+inline double chi_squared_cdf(double x, std::size_t degrees_of_freedom) {
```

Comment:
assert x and degrees_of_freedom non-negative?

Reply:
going to have it return NAN in that case to avoid putting an assert in the functions.